First Release of PSF Notebook: hst_point_spread_function

mrevalski opened the pull request:
This notebook checklist has been made available to us by the [Notebooks For All team](https://github.com/Iota-School/notebooks-for-all/blob/main/resources/event-hackathon/notebook-authoring-checklist.md).
Its purpose is to serve as a guide for both the notebook author and the technical reviewer highlighting critical aspects to consider when striving to develop an accessible and effective notebook.

### The First Cell

- [x] The title of the notebook in a first-level heading (eg. `<h1>` or `# in markdown`).
- [x] A brief description of the notebook.
- [x] A table of contents in an [ordered list](https://www.markdownguide.org/basic-syntax/#ordered-lists) (`1., 2.,` etc. in Markdown).
- [x] The author(s) and affiliation(s) (if relevant).
- [x] The date first published.
- [x] The date last edited (if relevant).
- [x] A link to the notebook's source(s) (if relevant).

### The Rest of the Cells

- [x] There is only one H1 (`#` in Markdown) used in the notebook.
- [x] The notebook uses other [heading tags](https://www.markdownguide.org/basic-syntax/#headings) in order (meaning it does not skip numbers). 

## Text

- [x] All link text is descriptive. It tells users where they will be taken if they open the link.
- [x] All acronyms are defined at least the first time they are used. 
- [x] Field-specific/specialized terms are used when needed, but not excessively.

## Code

- [x] Code sections are introduced and explained before they appear in the notebook. This can be fulfilled with a heading in a prior Markdown cell, a sentence preceding it, or a code comment in the code section.
- [x] Code has explanatory comments (if relevant). This is most important for long sections of code.
- [x] If the author has control over the syntax highlighting theme in the notebook, that theme has enough color contrast to be legible.
- [x] Code and code explanations focus on one task at a time. Unless comparison is the point of the notebook, only one method for completing the task is described at a time.

### Images

- [ ] All images (jpg, png, svgs) have an [image description](https://www.w3.org/WAI/tutorials/images/decision-tree/). This could be
    - [ ] [Alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt) (an `alt` property) 
    - [ ] [Empty alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt#decorative_images) for decorative images/images meant to be skipped (an `alt` attribute with no value)
    - [ ] Captions
    - [x] If no other options will work, the image is decribed in surrounding paragraphs.

- [ ] Any [text present in images](https://www.w3.org/WAI/WCAG21/Understanding/images-of-text.html) exists in a text form outside of the image (this can be alt text, captions, or surrounding text.)

### Visualizations

- [x] All visualizations have an image description. Review the previous section, Images, for more information on how to add it.
- [ ] [Visualization descriptions](http://diagramcenter.org/specific-guidelines-e.html) include
    - [ ] The type of visualization (like bar chart, scatter plot, etc.)
    - [ ] Title
    - [ ] Axis labels and range
    - [ ] Key or legend
    - [ ] An explanation of the visualization's significance to the notebook (like the trend, an outlier in the data, what the author learned from it, etc.)

- [ ] All visualizations and their parts have [enough color contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) ([color contrast checker](https://webaim.org/resources/contrastchecker/)) to be legible. Remember that transparent colors have lower contrast than their opaque versions.
- [ ] All visualizations [convey information with more visual cues than color coding](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html). Use text labels, patterns, or icons alongside color to achieve this.
- [ ] All visualizations have an additional way for notebook readers to access the information. Linking to the original data, including a table of the data in the same notebook, or sonifying the plot are all options.

mrevalski commented:
This draft pull request was created to run the DMD CI checks. Specifically, to understand if the FORTRAN executable call succeeds.

mrevalski commented:
Hi @FDauphin, I've corrected all of the PEP8 issues for this notebook and it is ready for your science review before Hatice completes the full technical review. @haticekaratay, I'm encountering an error with the execution checks, which state that a file is not found, but that file is produced in the previous cell via a FORTRAN executable in the same directory.

My initial thought is that I need to explicitly include the FORTRAN dependency in my requirements file, or perhaps it's not included in the CI system machinery, or simply that the architectures are different so the executable fails to run. As your time allows, do you have suggestions for how I should try to resolve this error? Thank you for your time and help!

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    options allow users to toggle diagnostic figures and 
+    messages on or off depending on the number of sources.
+    '''
+
```
For boolean variables, it's more Pythonic to write `if var` instead of `if var == True`
Change these for `verbose`, `scale_stars`, `sub_pixel`, and `show_figs`

mrevalski replied:
Hi Fred, thanks for these great suggestions. Just to clarify, you're suggesting to simply use the shorthand, e.g. `if verbose` rather than `if verbose is True`, is that correct? I like the later for readability as it's a reminder that the variable is boolean, but I understand that is probably explicit in the shorter version as well? Happy to change it.

FDauphin replied:
Hey Mitchell, correct I was suggesting using the shorthand. You're also right that explicitly stating can help with readability, but we also have the docstrings for that. However, I'm good keeping it; if it helps you it'll probably help another 🙂

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+        # Determine the location of the maximum flux.
+        peak_location = numpy.unravel_index(numpy.argmax(subimage, axis=None), subimage.shape)
+        if (verbose is True):
+            print('The subimage peak flux (x,y) = (' + str(peak_location[1]) + ', ' + str(peak_location[0]) + ')')
```
```
print(f'The subimage peak flux (x,y) = ({peak_location[1]}), {peak_location[0]})')
```

mrevalski replied:
Done.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+        if (peak_location[1] == 0 and peak_location[0] == 0):
+            print('This object is outside the data region and will be excluded.\n')
+            
+        if (peak_location[1] != 0 and peak_location[0] != 0):
```
Are there other sub conditions (e.g. `if peak_location[1] == 0 and peak_location[0] != 0`)? If not, using `else` should work here

mrevalski replied:
Unfortunately this is a bit complex. If all is well and the bright star is centered then the peak will be at (rpix, rpix). However, if the peak finder fails, it returns (0,0). However, it's also possible to have another bright star in the field so that the peak is not at (rpix, rpix), but the star can still be appended if a median is being taken. In this case I've combined them into an if-else statement, and users can modify if they'd like to reject sources with other bright stars.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    mysubplot[1].set_title(r'1\% Max')
+    mysubplot[2].set_title('LogNorm')
+    mysubplot[3].set_title('Log10')
+    for idx in [0, 1, 2, 3]:
```
`for idx in range(4)` is more Pythonic

mrevalski replied:
Done.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    if (show_figure is False):
+        matplotlib.pyplot.close()
+
+    return print('\nFigure saved as: '+filename)
```
Use f-string, also no need to return since it's not a variable that needs to be used later on
`print(f'\nFigure saved as: {filename}')`

mrevalski replied:
In that case, would it be best practice to return nothing, or some other form of status message?

FDauphin replied:
If you want `save_figure` to return a variable that will be used somewhere else, then return that. If not, it's best practice to simply print your statement and return nothing.

mrevalski replied:
Done.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    elif (stack_type == 'median'):
+        stacked_image = numpy.median(input_array, axis=0)
+    else:
+        return print('ERROR: stack_type must be mean or median.')
```
Raise error instead of printing
```
raise ValueError('stack_type must be mean or median.')
```

mrevalski replied:
Done.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    else:
+        return print('ERROR: stack_type must be mean or median.')
+
+    print('Creating a '+stack_type+' image stack.')
```
```
print(f'Creating a {stack_type} image stack.')
```

mrevalski replied:
Done.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    
+    figure = plot_cutouts(data=stacked_image, rpix=rpix)
+    
+    if (export_file != 'none'):
```
I might use an empty string as a default since if it's passed it might write towards the whole subdirectory (not sure though)

mrevalski replied:
Done. I've updated the docstring as well.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+        nrow = 1
+    elif (num_cutouts > max_cutouts):
+        nrow = int(max_cutouts / ncol)
+        print('Displaying a maximum of', max_cutouts, 'cutouts.')
```
```
print(f'Displaying a maximum of {max_cutouts} cutouts.')
```

mrevalski replied:
Done.

FDauphin commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    figure, mysubplot = matplotlib.pyplot.subplots(1, 4, figsize=(11, 11), sharex=True, sharey=True)
+    mysubplot[0].imshow(data, vmin=0.0, vmax=numpy.amax(data), origin='lower')
+    mysubplot[1].imshow(data, vmin=0.0, vmax=numpy.amax(data)/100.0, origin='lower')
+    mysubplot[2].imshow(data, norm=LogNorm(vmin=percentile(data, 0.0), vmax=percentile(data, 98.5)), origin='lower', aspect='equal')
```
`np.percentile`

mrevalski replied:
Done.

FDauphin reviewed (changes requested):
Hey @mrevalski , great start on this! Your coding style is very clean and logical. I'll only be reviewing `psf_utilities` and the notebook since that's where the bulk of changes need to be made. I will be working on the notebook later this afternoon! Here are some overall notes:
- Use np and plt for numpy and matplotlib as that's standard practice
- Use the Numpy notation for docstrings which includes function, parameter, and return definitions
- Use f-strings when possible to condense some print statements
- For boolean variables, it's standard to write `if var` instead of `if var == True`

Feel free to let me know if you have any questions; awesome job again 👍🏾

mrevalski commented on notebooks/WFC3/point_spread_function.md:
```diff
@@ -0,0 +1,12 @@
+## HST Point Spread Function
+This notebook demonstrates how to generate Point Spread Function (PSF) 
+models for WFC3 observations. This includes retrieving empirical models 
+from the [WFC3 PSF Website](https://www.google.com/search?client=safari&rls=en&q=wfc3+psf&ie=UTF-8&oe=UTF-8), generating custom models by stacking stars, 
```
Corrected.

FDauphin reviewed (changes requested):
Hey @mrevalski, this is really an exceptional notebook! You did a great job explaining each method and kept the code relatively succinct. Also all of your plots are clear and easy to understand. Here are some overall comments:

- ISRs are formally stated as WFC3 ISR YYYY-XX
- Make sure your imports in markdown and code accurately reflect each other
- Remove any imports, commended code, or objects that aren't being used
- Adding some more markdown on isolated code cells will add more context for the user (usually don't want two code cells back to back)
- Save your diagram in the repo and call that file in the markdown cell. It'll remove a massive line of text to generate the image and make the raw ipynb file a lot smaller
- It would be nice to save the perturbation results separately instead of overwriting
- Restating the various hyperlink sources in the Additional Resources section will be helpful

A good amount of comments, but overall excellent job. Was also able to run everything with no errors (including hst1pass) which is a huge plus 🙌🏾

mrevalski commented:
> Hi @FDauphin, I've corrected all of the PEP8 issues for this notebook and it is ready for your science review before Hatice completes the full technical review. @haticekaratay, I'm encountering an error with the execution checks, which state that a file is not found, but that file is produced in the previous cell via a FORTRAN executable in the same directory.
> 
> My initial thought is that I need to explicitly include the FORTRAN dependency in my requirements file, or perhaps it's not included in the CI system machinery, or simply that the architectures are different so the executable fails to run. As your time allows, do you have suggestions for how I should try to resolve this error? Thank you for your time and help!

Hi everyone, Fred Dauphin was able to help me solve this issue. The solution was two fold. First, we added a pre-requirements.sh file with "conda install --yes -c conda-forge gfortran" to ensure FORTRAN was installed in the CI environment. Next, we removed the pre-compiled executable, provided the FORTRAN code in the form of a *.F source code file, and then had the notebook compile it with a BASH command. The full set of CI checks now pass!

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+import os
+import shutil
+import urllib
+import numpy
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+import shutil
+import urllib
+import numpy
+from numpy import percentile
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+import numpy
+from numpy import percentile
+import matplotlib
+import matplotlib.pyplot
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    mask[0:ycenter-int(cutout_size/2), 0:numpy.shape(data)[1]] = True
+    mask[0:numpy.shape(data)[0], 0:xcenter-int(cutout_size/2)] = True
+    mask[0:numpy.shape(data)[0], xcenter+int(cutout_size/2):numpy.shape(data)[1]] = True
+
```
I think this may have the opposite effect. Specifically, each of the four lines is setting the top, bottom, left, and right regions _outside_ of the box to be True, while the refactored version is setting the values _inside_ the box to True?

Alternatively, I could set the entire array to True, and then use the line you provided to set the inside to False?

FDauphin replied:
Oh I see, then the alternative version would be:
```
mask = np.ones(data.shape, dtype=bool)
size = int(cutout_size / 2)
mask[ycenter-size:ycenter+size, xcenter-size:xcenter+size] = False
```
Up to you if you want to update the coding block 😃

mrevalski replied:
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+def plot_apertures(data, xcenter, ycenter, cutout_size, apertures_stellar, apertures_annulus):
+
+    '''
+    A function that takes an image, x and y center locations, image size, and 
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    my_figure_size, my_fontsize = setup_matplotlib('notebook', 1.2)
+    figure, axes = matplotlib.pyplot.subplots(1, 3, figsize=(11.0, 11.0/3))
+    # current_width = 1.2 * matplotlib.pyplot.rcParams["lines.linewidth"]
+    for ax in [0, 1, 2]:
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    axes[2].set_title('Photometry Apertures')
+    axes[0].set_ylabel('Y (pixels)')
+    axes[1].set_xlabel('X (pixels)')
+    for ax in [0, 1, 2]: 
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    # Create a figure with the data, identified stars, and photometry apertures.
+    my_figure_size, my_fontsize = setup_matplotlib('notebook', 1.2)
+    figure, axes = matplotlib.pyplot.subplots(1, 3, figsize=(11.0, 11.0/3))
+    # current_width = 1.2 * matplotlib.pyplot.rcParams["lines.linewidth"]
```
Done and done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    # Make a figure showing the data, model, and residuals.
+    my_figure_size, my_fontsize = setup_matplotlib('notebook', 1.2)
+    figure, axes = matplotlib.pyplot.subplots(1, 3, figsize=(11.0, 11.0/3))
+    # current_width = 1.2 * matplotlib.pyplot.rcParams["lines.linewidth"]
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+        star_id = star_ids[i]
+        
+        # Print the (x, y) coordinates and extract each star image.
+        print('Star ID ' + str(int(star_id)) + ':' + ' (x,y) = (' + str(xi) + ', ' + str(yi) + ')')
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    axes[0].set_title('Data')
+    axes[1].set_title('Model')
+    axes[2].set_title('Residuals')
+    for x in [0, 1, 2]: 
```
In another function I loop over it because the science data is being shown in all panels. I'll keep it here for simplicity.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    Download a PSF model from the WFC3 website and validate the file.
+    '''
+
+    if (detector not in ['WFC3UV', 'WFC3IR', 'ACSWFC']):
```
In this case the function is downloading Jay Anderson's `PSFSTD` formatted models, which I do not believe are available for WFPC2. In fact, I'm unable to locate the models for ACS, which will cause a crash. I've contacted Jay.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    if (detector not in ['WFC3UV', 'WFC3IR', 'ACSWFC']):
+        print('The valid detector options are: WFC3UV, WFC3IR, or ACSWFC')
+
+    psf_name = 'PSFSTD_{}_{}.fits'.format(detector, filter)
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Corrected, thank you.

mrevalski replied:
Okay, that is strange, I've corrected it and confirmed the fix online.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "\n",
+    "***\n",
+    "\n",
+    "The following Python packages are required to run the Jupyter Notebook:\n",
```
Done. These are required in psf_utilities.py and so have been retained in the markdown description of the imports.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "     - [detection.DAOStarFinder](https://photutils.readthedocs.io/en/stable/api/photutils.detection.DAOStarFinder.html#photutils.detection.DAOStarFinder) - finding stars in images\n",
+    "     - [detection.find_peaks](https://photutils.readthedocs.io/en/stable/api/photutils.detection.find_peaks.html#photutils.detection.find_peaks) - finding peaks in images\n",
+    "     - [photutils.centroids](https://photutils.readthedocs.io/en/stable/centroids.html) - finding stellar centroids\n",
+    "     - [photutils.aperture](https://photutils.readthedocs.io/en/stable/aperture.html) - aperture photometry tools\n",
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
```
I've removed `matplotlib`. `Table` is used to read the tweakreg files, source catalogs, etc.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "import shutil\n",
+    "import tarfile\n",
+    "from IPython.display import clear_output\n",
+    "import numpy\n",
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "from IPython.display import clear_output\n",
+    "import numpy\n",
+    "import matplotlib\n",
+    "import matplotlib.pyplot\n",
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "     - [modeling.fitting](https://docs.astropy.org/en/stable/modeling/fitting.html) - fitting methods for models\n",
+    "     - [visualization.simple_norm](https://docs.astropy.org/en/stable/api/astropy.visualization.mpl_normalize.simple_norm.html) - display image normalization\n",
+    "     - [stats.sigma_clipped_stats](https://docs.astropy.org/en/stable/api/astropy.stats.sigma_clipped_stats.html) - mean, median, and standard deviation\n",
+    "     - [stats.sigma_clip](https://docs.astropy.org/en/stable/api/astropy.stats.sigma_clip.html) - sigma-clip provided data\n",
```
Corrected to `SigmaClip`, thank you.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "\n",
+    "**What is the Point Spread Function and why do we want to model it?**\n",
+    "\n",
+    "The purpose of a telescope is to collect, magnify, and record the intensity of light from astronomical sources. A telescope's ability to to resolve the size of objects is limited by its spatial resolution, which is determined by its aperture and the wavelength of light being observed, such that every telescope is fundamentally limited by the diffraction nature of light. The minimum size of a ...
```
Thank you! 😊

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+   "id": "316b41c6-fd8d-4218-a548-0cdbb8b84500",
+   "metadata": {},
+   "source": [
+    "Set the main code, data, PSF, and plot directories. Set the backend configuration option to 'retina', which greatly improves the resolution of figures in notebooks."
```
Corrected, thanks.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "               apertures_stellar=apertures_stellar, \n",
+    "               apertures_annulus=apertures_annulus)\n",
+    "\n",
+    "save_figure(figure, plot_dir+'fig1.pdf', True)"
```
In my experience, PDFs effectively have "infinite" resolution, while JPEGs and PNGs get blurry when zoomed in on. Thoughts?

FDauphin replied:
Did not know that; sounds good to me 😄

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst1pass_darwin.e:
Agreed, this has been fixed now that the hst1pass.F source code file is included, and compiled within the notebook. I have removed this file from the repository.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
No, this was a temporary issue while I was sorting out latex issues. I have latex on my system but it crashed on the CI, so I either need to have latex as a pre-requirement for the CI, or more simply as I've done here, just use normal text. Thoughts?

FDauphin replied:
Just using the normal text is fine since the titles are still clear.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
This is a great reason to ensure that the various results are saved with different filenames. I will fix this!

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
````diff
@@ -0,0 +1,1404 @@
+    "\n",
+    "FORTRAN can be installed by typing the command ```conda install -c conda-forge gfortran``` into a terminal. If the installation is successful, then typing ```gfortran --version``` should return a message similar to the following: ```GNU Fortran (GCC) 13.2.0 Copyright (C) 2023 Free Software Foundation, Inc.```.\n",
+    "\n",
+    "Next, to install hst1pass proceed to [https://www.stsci.edu/~jayander/HST1PASS/CODE/hst1pass/](https://www.stsci.edu/~jayander/HST1PASS/CODE/hst1pass/) and download the latest version of ```hst1pass.F```, e.g.: ```hst1pass.2023.11.07_v1f.F```. On MacOS, save this to your computer as ```hst1pass_darwin.F```. Finally, compile the code using the command: ```gfortran hst1pass_darwin.F -o hst1pas ...
````
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+from photutils.centroids import centroid_com
+from photutils.aperture import CircularAperture
+from scipy.ndimage import shift
+
```
It's interesting that the industry still follows standards developed from punch cards 😄, but I do get that this helps with readability and even some terminal applications. I've adjusted the line length where easy, but allowed a few lines (e.g. list of function parameters) to extend beyond this. Just let me know if you'd like me to be stricter with this limit.

FDauphin replied:
I know funny how legacy persists 😆 Everything looks good to me 👍🏾

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    
+    # Create a mask array that is the same size as the science image and mask the upper, lower, left, and right regions.
+    mask = numpy.zeros(data.shape, dtype=bool)
+    mask[ycenter+int(cutout_size/2):numpy.shape(data)[0], 0:numpy.shape(data)[1]] = True
```
Because python reverses standard (x, y) row, column notation I find it helpful to program it explicitly.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    return my_figure_size, my_fontsize
+
+
+def create_mask(data, cutout_size, xcenter, ycenter):
```
In a few cases within the notebook the variable is used directly and rpix has a different definition so opting to keep it.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    psf_path = '{}/{}'.format(file_path, psf_name)
+    psf_url = 'https://www.stsci.edu/files/live/sites/www/files/home/hst/instrumentation/wfc3/data-analysis/psf/_documents/'
+
+    # Download the PSF file if it doesn't exist.
```
Originally I thought that hst1pass required the PSF models in the same directory as the executable, but it turns out that in completing the file validation check the function was looking for the file rather than the filepath. I've updated this and now the PSF models, both from STScI and those produced by the user, are only saved once in the `psfs_dir`. Thanks!

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+    for i in range(len(xis)):
+        
+        # Convert to integer and round to nearest value.
+        xi = int(numpy.rint(xis[i]))
```
Hi Fred, thanks for confirming this important point. Indeed, this is the desired behavior because we always want stars that are offset by up to and including half a pixel to be rounded to the same integer array value so they will be properly centered and aligned at the subpixel level. I confirmed the rounding is going the right direction by also calculating the center of mass after the shift, and the centroids (from hst1pass or photuils in these cases) always agree to within better than 0.05 pixels. However, I did not explicitly test with a centroid that is an exact half pixel value (e.g. 19.500000). Considering that hst1pass and photuils centroid to several decimal places this would seem rare, but still important. I think all is well and we can proceed with the code as is in this case.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+            if (verbose is True):
+                print('x_shift, y_shift =', x_shift, y_shift)
+
+            '''
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,402 @@
+            if (verbose is True):
+                print('Warning: The function is not aligning at the sub-pixel level. Is this intended?')
+
+        # # If desired we can recentroid. This would be false using hst1pass and true for approximate centroids.
```
Agreed, this was mostly for testing, and could have some use cases, but it's simpler to have users supply exact coordinates. They could always use the center of mass to calculate the centroids and provide those to the function. I have removed the extraneous commented out code.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Great idea! Would you also suggest moving the figure, *.py, and FORTRAN files to a "docs" folder, or simply keep them in the main directory? I think the FORTRAN files need to stay at the main level, so perhaps it's fine with these few files.

FDauphin replied:
I think this repo is comfortably at its (soft) limit of files at the top level so I'm good with either. I can see `docs` containing the modules, the hst1pass output, and the workflow image, but "docs" doesn't seem like the best name nor can I think of a good name for a folder with those files 😅

mrevalski replied:
Agreed, I'll leave everything at the main level for now, and we can revisit in the future if the repo grows.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
I've updated the text to highlight that it's the left fork on the right branch, and that the function is within wfc3tools:

Note that the left fork under the drizzle level branch is an advanced workflow that involves drizzling the empirical models provided by the WFC3 team into FLC files. This requires installing a custom version of [wfc3tools](https://github.com/Vb2341/wfc3tools), running the make_model_star_image() function within wfc3tools, and then stacking the drizzled models. This functionality is under development and not yet officially supported.

FDauphin replied:
Thanks. Could you directly link to the function or be more explicit as to where to find it? I see you linked Varun's fork, but is that function in the master branch or another? What module does it exist in (e.g. `utils.py`)? I'm having a hard time getting there myself, but maybe I'm missing something in plain sight 🤦🏾‍♂️

mrevalski replied:
The reason you could not find it is because somehow I linked wfc3tools instead of wfc3_photometry 🙃 

I have updated the link to: [wfc3_photometry](https://github.com/spacetelescope/wfc3_photometry).

The text now reads:

"This requires installing the [wfc3_photometry](https://github.com/spacetelescope/wfc3_photometry) package, running the make_model_star_image() function within PSFUtils.py, and then stacking the drizzled models. This functionality is under development and not yet officially supported."

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Possibly, but having variables for fit_shape, min_separation, threshold, fwhm, roundi, and aperture_radius seemed like a lot, and the user will likely wan't to change these frequently, so rather than having them look at the functions in psf_utilities, I thought it may be easier for them to see how they're defined directly?

FDauphin replied:
Yeah that sounds good 👍🏾

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
I've greatly expanded the explanation and what users should look for in the results:

In this first analytical example we fit Gaussian profiles to the four stars in our subimage and display the data, model, and residuals. We provide an initial guess for the width (sigma) using an approximate full width at half maximum (FWHM) of 2 pixels, where $\sigma$ = FWHM/2.355. Users are encouraged to experiment with this value to see how the resulting fit and residuals change. First, the `calculate_photometry` function uses the provided PSF model with the photutils `PSFPhotometry` function to fit the model to the stars. In essense, the flux of the normalized PSF model is scaled up to match the flux of each star. In this case, neighboring stars are fit simultanously. The `make_residual_image` function then subtracts the best fit model so the flux residuals can be examined. Finally, we call the `calculate_residuals` function with the exact same stellar and background photometry apertures used in Section 1.4 in order to measure the residual fluxes. A smaller residual flux generally implies a better fit, but the residuals should be inspected as over and undersubtractions from different portions of the star can also yield a small residual flux.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Yes, I've clarified this by adding:

The three panels above are shown with the same minimum and maximum flux values, so the residuals can be directly compared with the data and model.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Interestingly, this is how it's written in the documentation, perhaps some circles refer to chips 1 and 2 as WFC2 and WFC1, respectively? See: https://photutils.readthedocs.io/en/stable/api/photutils.psf.stdpsf_reader.html

Regardless, I've clarified it using more standard nomenclature as suggested.

FDauphin replied:
I misread that as just for WFC3, but I see its also for ACS, in which case what you had before was correct 🙃 🤦🏾‍♂️ apologies about that feel free to revert to the original acronyms

mrevalski replied:
Reverted!

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
The Github preview above doesn't show any code or text, just a green box with nonsense raw code, but I'm guessing that the hyperlink should go to MAST where is says "centroids provided by the MAST cutout service", is the correct?

FDauphin replied:
Good guess; thanks! However, I'm not sure what's going on with the GitHub preview ...

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done. Thanks, this makes the code more concise!

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
I've commented it out and added an in-line comment that it would be used for UVIS2, is that acceptable? I want to be sure that users see this in each of 2.2.1, 2.2.2, 2.2.3 so they don't forget to use the correct setup for their data. Thoughts?

FDauphin replied:
Yeah that's a good idea 👍🏾

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Addressed above.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Corrected.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Okay I see the logic and agree, I've made the change and ordered the parameter list in the same order as the `mast_api_psf.set_min_max` listings.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "\n",
+    "The purpose of a telescope is to collect, magnify, and record the intensity of light from astronomical sources. A telescope's ability to to resolve the size of objects is limited by its spatial resolution, which is determined by its aperture and the wavelength of light being observed, such that every telescope is fundamentally limited by the diffraction nature of light. The minimum size of a ...
+    "\n",
+    "While a well-focused telescope produces a very small image of point sources, the shape will be distorted and the light is effectively scattered to an infinite radius. Therefore, a model of the PSF is required to accurately measure astrometry (positions) and photometry (brightnesses) of objects in astronomical images. In the case of the HST WFC3 this is especially important because the images ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "\n",
+    "Finally, astrometric and photometric measurements can generally be made either from individual calibrated exposures known as [FLTs and FLCs](https://hst-docs.stsci.edu/wfc3dhb/chapter-2-wfc3-data-structure/2-1-types-of-wfc3-files) for WFC3, or exposures that are combined into a mosaic through drizzling. In theory, modeling individual exposures is always preferable because these images are on ...
+    "\n",
+    "Users looking for a fundamental introduction to [aperture photometry](https://boyce-astro.org/aperture-photometry/) and [PSF photometry](https://boyce-astro.org/point-spread-function-psf/) may find this excellent series of [video tutorials](https://boyce-astro.org/videos/photometry/) by Boyce-Astro helpful, as well as this description of the [Airy Disk](https://www.edmundoptics.com/knowledge ...
```
Opted to keep in both locations for completeness.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "\n",
+    "<div class=\"alert alert-block alert-info\"><b>NOTE:</b> The following non-Python tools are needed to run some advanced features of the Jupyter Notebook. Specifically, a FORTRAN compiler and the <code>hst1pass</code> FORTRAN code are only required for PSF perturbation with individual exposures (FLTs/FLCs). These tools are not required for stacking stars or creating PSF models for drizzled im ...
+    "\n",
+    " - [**hst1pass**](https://www.stsci.edu/~jayander/HST1PASS/) - FORTRAN code for performing photometry and PSF perturbation as documented in [Anderson 2022 (WFC3 2022-5)](https://www.stsci.edu/files/live/sites/www/files/home/hst/instrumentation/wfc3/documentation/instrument-science-reports-isrs/_documents/2022/WFC3-ISR-2022-05.pdf).\n",
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "from astropy.stats import sigma_clipped_stats\n",
+    "from astropy.stats import SigmaClip\n",
+    "from photutils.detection import DAOStarFinder\n",
+    "from photutils.aperture import aperture_photometry\n",
```
Done. I applied the same logic to the astropy.stats and drizzlepac modules.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "from astropy.table import Table, QTable\n",
+    "from astropy.visualization import simple_norm\n",
+    "from astropy.stats import sigma_clipped_stats\n",
+    "from astropy.stats import SigmaClip\n",
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "from drizzlepac import tweakreg\n",
+    "from drizzlepac import astrodrizzle\n",
+    "\n",
+    "# Custom functions written in psf_utilities.py.\n",
```
I agree this would condense the import statements, but then requires calling e.g. `psf_utilities.create_mask`, which lengthens the function calls throughout the full notebook?

FDauphin replied:
True, might be easier to just keep as is 👍🏾

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+    "figure, axes = matplotlib.pyplot.subplots(1, 1, figsize=my_figure_size)\n",
+    "cutout_size = 75 # 75 WFC3 UVIS pixels = 3 arcseconds.\n",
+    "\n",
+    "axes.add_patch(Rectangle((1818-cutout_size/2, 986-cutout_size/2), width=cutout_size, height=cutout_size, fill=None, color='w'))\n",
```
In an (admittedly small) number of cases the full variable is used, and provides the user with a more intuitive way to select the size of the cutout, allowing the code to do the array sizing in the background. Opting to keep it for now.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Okay I've added a yellow warning cell that states: "NOTE: Users who are unable to run hst1pass can execute the tutorial by copying the catalog file id8048dyq_flc.xympqks stored in this repository to their data_dir." I could add a cell that tries to copy it for them if it doesn't exist, but that may be overkill?

FDauphin replied:
I think the NOTE cell should be enough, nice addition 👍🏾

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Okay, users can go to the repository link at the top. Since this is intended for new users, I've included both links anyway.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
I've added a link to the astrodrizzle readthedocs that describes each parameter and omitted more detailed explanations to keep the notebook concise. There are in-line comments for each parameter that hopefully provide some clarity.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Corrected.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented:
Hi @FDauphin, thank you for the very helpful review! Your revisions have greatly improved the clarity and coding cleanliness of the notebook. I've incorporated the vast majority of your comments exactly as suggested, and replied to a small number above for discussion. I think the main outstanding items are the following:

- Should we move all of the auxiliary files such as the workflow diagram, catalog and FORTRAN files, to a "docs" folder?
- Is it okay to use `if var is True` for Boolean statements, or should I abbreviate this to simply `if var` for conciseness?
- It would be beneficial to add a comparison of all the methods, but this is tricky as I don't want to give the impression that any one method is "better" as it depends on the dataset and science goals. Similarly the drizzle case is separate. Thoughts?

In keeping @haticekaratay in the loop, once Fred has reviewed my changes and we iterate on the last few items listed above, we'll convert this from a draft to a full pull request for your review. We hope to give you a full working week to review and revise this with us, with the goal of pulling it into the main branch before June 7 so we can link to the new notebook in the Space Telescope Announcement Newsletter (STAN). Please let me know if this timeline will work with your existing commitments.

Have a great day!

FDauphin commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1531 @@
+    "     - [detection.find_peaks](https://photutils.readthedocs.io/en/stable/api/photutils.detection.find_peaks.html#photutils.detection.find_peaks) - finding peaks in images\n",
+    "     - [photutils.centroids](https://photutils.readthedocs.io/en/stable/centroids.html) - finding stellar centroids\n",
+    "     - [photutils.aperture](https://photutils.readthedocs.io/en/stable/aperture.html) - aperture photometry tools\n",
+    "       - [aperture_photometry](https://photutils.readthedocs.io/en/stable/aperture.html) - measure aperture photometry\n",
```
Correct [link](https://photutils.readthedocs.io/en/stable/api/photutils.aperture.aperture_photometry.html#photutils.aperture.aperture_photometry) for `aperture_photometry()`

mrevalski replied:
Corrected!

FDauphin commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1531 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "os.chdir(data_dir)\n",
```
Maybe instead of rerunning hst1pass (which could take some time depending on others computational resources), you could make save a copy of each pert results then overwrite the pert1 results to the desired name:
```
# after running hst1pass in 2.2.2
shutil.copy('id8048dyq_flc.xympqks', 'id8048dyq_flc_pert1.xympqks')
# after running hst1pass in 2.2.3 (for consistency)
shutil.copy('id8048dyq_flc.xympqks', 'id8048dyq_flc_pert2.xympqks')
# after plotting pert2 results in 2.2.3
shutil.copy('id8048dyq_flc_pert1.xympqks', 'id8048dyq_flc.xympqks')
```

mrevalski replied:
Okay, I've taken your advice here with a single tweak so users can more easily change catalogs, and so the main catalog is copied to the main directory for the repository so it's updated automatically. I updated the text:

"Finally, we copy the hst1pass catalog generated with PERT=1 so the best catalog is used later in the stacking analysis. We also copy the catalog to the main directory so users who are unable to run hst1pass can complete the tutorial. Users can change this cell to experiment with different catalogs made by running hst1pass with different settings."

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+    A function that displays a grid of MAST PSF cutouts that have 
+    been retrieved using the functions from mast_api_psf.py written 
+    by Fred Dauphin. The function accepts a list of 2D arrays 
+    corresponding to cutouts and generates a gris plot with a 
```
Typo corrected.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+
+    """
+    A function that displays a grid of MAST PSF cutouts that have 
+    been retrieved using the functions from mast_api_psf.py written 
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+
+    if (max_cutouts > len(cutouts)):
+        max_cutouts = ((len(cutouts) % 5 + 5)*5)
+        print('Setting max_cutouts to', max_cutouts, 'so it is less than the number of', len(cutouts), 'cutouts.')
```
Done.

FDauphin reviewed (changes requested):
Thanks for addressing each comment, everything is coming together nicely! All of your functions are clearly explained. There's some small documentation changes and f-string suggestions. I also replied to a few comments, including:

- psf_utilites: return for `save_figure()`
- psf_utilities: optimizing `create_mask()`
- notebook: making `docs` folder
- notebook: `make_model_star_image` in `wfc3tools`
- notebook: WFC/UVIS for ACS/WFC3
- notebook: `# was flt`
- notebook: `numpy` citation

In addition, I think adding more descriptive filenames to your figures will be helpful. It was hard to remember what "fig6" was, and it's nice seeing all the titles if several of them are open in Adobe.

After these get addressed, we should be good for a DMD review 🙌🏾

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+        A list of the y-centroids of each star, assumed to be exact.
+    rpix : int
+        An integer number of pixels for the cutout length and width.
+    scale_stars : bool
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+
+    psf_name = f'STDPSF_{detector}_{filter}.fits'
+    psf_path = f'{file_path}{psf_name}'
+    psf_url = 'https://www.stsci.edu/~jayander/HST1PASS/LIB/PSFs/STDPSFs/'+detector+'/'
```
Corrected!

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+def plot_apertures(data, xcenter, ycenter, cutout_size, apertures_stellar, apertures_annulus):
+
+    """
+    A function that takes an image, x and y center locations, cutout size, and 
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+
+    """
+    Given an array containing 2D arrays with centered images of stars, 
+    create and mean or median stack. The 'rpix' parameter sets the size 
```
Typo corrected.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+        after being cropped to a subarray using the make_cutouts() function.
+    rpix : int
+        An integer number of pixels for the cutout length and width.
+    stack_type : str
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+        The mathematical averaging function, either median or mean. 
+        In general, median is recommended for rejecting contaminants.
+    scale_flux : bool
+        If True then the value of each pixel is divided by the maximum. 
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+    max_cutouts : int
+        The maximum number of cutouts to display on screen. The code 
+        will adapt, but intervals of five make for easier displaying.
+    verbose : bool
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1531 @@
+    "\n",
+    "***\n",
+    "\n",
+    "The example above demonstrates how a Gaussian profile can determine reasonable centroids for stellar sources, but they fail to accurately model the radial light distribution. While it is possible to use more advanced analytical forms such as [Moffat profiles](https://docs.astropy.org/en/stable/api/astropy.modeling.functional_models.Moffat2D.html), the overall shape can be more accurately rep ...
```
Corrected.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1531 @@
+   "id": "7dc383b4-a3fa-45bb-b116-91049b317bd6",
+   "metadata": {},
+   "source": [
+    "Thank you for completing this tutorial."
```
Annalisa noted that most of the information in the conclusions, such as pointing out that this can also be used with ACS, etc. really would be more appropriate in the introduction, so I've moved that information there. A formal conclusions section seemed a bit superfluous, although this could be a great spot to incorporate the summary you suggested.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
Done.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
I agree that a detailed comparison may need to wait for version 2.0. In this case, users can open all of the figures side-by-side to compare, as well as compare the residual flux measurements printed in the notebook. Will save for later.

mrevalski commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1404 @@
+  {
+   "attachments": {
+    "574e14a9-97d1-4e3f-b4da-3c09b5c8946a.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAACvAAAAgiCAYAAACw40JVAAAKrmlDQ1BJQ0MgUHJvZmlsZQAASImVlwdUU+kSgOfe9EZLiICU0JsgnQBSQmgBFKSDjZAECCXGUFTsyOIKrgUVESwruiqiYKOIHQu2RbH3BVkE1HWxYEPlXeAQdved9955c86c+e7c+Wfmv+f/z5kLwGAJ5fIMVA0gU5atiAj05cTFJ3BIPYADGlBgAhCFoiw5Lzw8FDAZtX+XD3cBGbK3bIZy/fv7/yrqYkmWCAAJxzhJnCXKxPgopn0iuSIbALcL8xvPzZYP8SWMWQqsQYwfD3HKCPcNcdIw4/HDMVERfIy1AMh0oVCRAkA3wfycXFEKlofuh7GdTCyV ...
```
I agree that this could be very useful for users, although I'm hesitant to go into too much detail because the "best" method will heavily depend on their dataset and goals (e.g. they might accept a worse qfit to capture wing emission, etc.). I'm hoping the descriptions for each method that weigh the pro's and con's of each approach are sufficient, at least until the next iteration.

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+        'WFPC2', 'ACSWFC', 'ACSSBC', and 'ACSHRC'.
+    filter : str
+        The HST photometric filter, valid for all filters with empirical 
+        models as listed on the WFC3 PSF webpage. Models are generally 
```
I've also added a direct link to Jay's library in the docstring:

https://www.stsci.edu/~jayander/HST1PASS/LIB/PSFs/STDPSFs/

mrevalski commented on notebooks/WFC3/point_spread_function/psf_utilities.py:
```diff
@@ -0,0 +1,607 @@
+
+    Returns
+    -------
+    image_array : np.ndarray
```
Interestingly, the stack_cutouts function works with either a list or array input. I think in this case because it's providing a list of arrays to take the median of in stack_cutouts. I've added the suggested line of code in make_cutouts.

mrevalski commented:
> Thanks for addressing each comment, everything is coming together nicely! All of your functions are clearly explained. There's some small documentation changes and f-string suggestions. I also replied to a few comments, including:
> 
> * psf_utilites: return for `save_figure()`
> * psf_utilities: optimizing `create_mask()`
> * notebook: making `docs` folder
> * notebook: `make_model_star_image` in `wfc3tools`
> * notebook: WFC/UVIS for ACS/WFC3
> * notebook: `# was flt`
> * notebook: `numpy` citation
> 
> In addition, I think adding more descriptive filenames to your figures will be helpful. It was hard to remember what "fig6" was, and it's nice seeing all the titles if several of them are open in Adobe.
> 
> After these get addressed, we should be good for a DMD review 🙌🏾

Hi @FDauphin, thank you for the detailed and thoughtful review! I have incorporated all of your comments to the best of my ability, and the notebook is ready for your final review before we pass it to Hatice. Please let me know if you see any issues. In regards to `image_array` I updated this to `image_list` and fixed the docstring, as otherwise it causes a crash.

As a final question, do you think the descriptive list of imports is overly verbose? For example, I could remove the third level of bullets under `photutils.aperture` and `photutils.psf` and simply link to the main packages. It doesn't hurt since the information is already there, but it may help to make the notebook more concise?

I think we have a nice list of features that would be excellent to add in "v2.0", which I'll tackle in the future after discussions with Sylvia and the PSF team. Please let me know when you've completed your review and I'll convert this from a draft to a formal pull request. Thank you again, I really appreciate your time and help! 🙂

FDauphin commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1542 @@
+   "outputs": [],
+   "source": [
+    "shutil.copy(data_dir+'id8048dyq_flc_pert1.xympqks', data_dir+'id8048dyq_flc.xympqks') # Used below for source selection.\n",
+    "shutil.copy(data_dir+'id8048dyq_flc_pert1.xympqks', code_dir+'id8048dyq_flc.xympqks') # Used for the Github repository."
```
If the second line is for non-Fortran users, there wouldn't be a `pert1` file right? Would this resolution copy the file in the reop to the data directory:
```
shutil.copy(code_dir+'id8048dyq_flc.xympqks', data_dir+'id8048dyq_flc.xympqks')
```

mrevalski replied:
Hi Fred, good logic error catch! I think the ideal behavior would be to copy the PERT1 catalog if it exists (meaning users ran hst1pass), otherwise copy it from the main folder in the repo to the data directory. How's this?

`if os.path.exists(data_dir+'id8048dyq_flc_pert1.xympqks'):
    shutil.copy(data_dir+'id8048dyq_flc_pert1.xympqks', data_dir+'id8048dyq_flc.xympqks') # Used below for source selection.
else:
    shutil.copy(code_dir+'id8048dyq_flc.xympqks', data_dir+'id8048dyq_flc.xympqks') # Included in the Github repository.`

FDauphin replied:
Yeah that'll work, thanks for cleaning that up 🙌🏾

FDauphin reviewed (commented):
Hey @mrevalski thanks for all your hard work! Replying to your comments here:

- A possible bug on copying output files.
- If you want to remove the third bulleted layer for `photutils` (aperture and psf), that's fine with me. The original problem was one of them went out to the third layer and the other didn't. As long as it's consistent, either to two or three layers is fine.
- For conclusions, it's mostly for consistency across WFC3 notebooks, as each one has a small conclusions section. There you can quickly bullet point the various methods performed and learned. 

I'm confident these will be your final commits. Thanks for being patient!

mrevalski commented:
> Hey @mrevalski thanks for all your hard work! Replying to your comments here:
> 
> * A possible bug on copying pert output files.
> * If you want to remove the third bulleted layer for `photutils` (aperture and psf), that's fine with me. The original problem was one of them went out to the third layer and the other didn't. As long as it's consistent, either to two or three layers is fine.
> * For conclusions section, it's mostly for consistency across WFC3 notebooks, as each one has a small conclusions section. There you can quickly bullet point the various methods performed and learned.
> 
> I'm confident these will be your final commits. Thanks for being patient!

Hi @FDauphin, thank you for the quick response and excellent feedback. I've corrected the catalog copying error, and decided to retain the third level of bullets for now. If more packages are used in the future then we'll abbreviate this list. Great idea, I've added back in the Conclusions section, and created a brief summary of all the techniques, linked to each section, for the user.

Many thanks again! Please let me know if you have any remaining suggestions, or we're ready for DMD review. 🙂

FDauphin reviewed (approved):
@mrevalski great work; looks like we're good to proceed with the DMD review 👍🏾 
@haticekaratay this PR is ready whenever you are

haticekaratay commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1551 @@
+{
```
<p>Let's group these imports from the same module together:</p>

<pre class="ql-syntax" spellcheck="false">from psf_utilities import ( save_figure, setup_matplotlib, ...
                           ....)
</pre><p> </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/hst_notebooks/pull/270/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

mrevalski replied:
I agree and I have implemented this change. I have also split the imports onto a new line using the "\" continuation character so that the full list of imported functions is visible without scrolling horizontally in the notebook cell. Please let me know if that is an issue.

haticekaratay commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1551 @@
+{
```
<p>Your notebook is well-structured and clear. For directory creation logic, I recommend using <code>if not os.path.exists(folder):</code> to check dir before creating it. This makes the intent more explicit.</p><p> </p>

<pre class="ql-syntax" spellcheck="false">for folder in ['data', 'psfs', 'plot', 'driz', 'tweak']:
&nbsp; &nbsp; if not os.path.exists(folder):
&nbsp; &nbsp; &nbsp; &nbsp; os.mkdir(folder)
&nbsp; &nbsp; else:
&nbsp; &nbsp; &nbsp; &nbsp; print(f'{folder} directory exists.')
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/hst_notebooks/pull/270/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

mrevalski replied:
Agreed and implemented.

haticekaratay commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1551 @@
+{
```
<p>Great work, just a minor suggestion:</p><p>I recommend use of context manager when working with FITS files.</p>

<pre class="ql-syntax" spellcheck="false">with fits.open(data_dir+hst_wfc3_f606w_file) as hdul:
    sci_data = hdul['SCI', 1].data
print('The image dimensions are:', np.shape(sci_data))
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/hst_notebooks/pull/270/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

haticekaratay replied:
<p>Let's use a context manager here also.</p>

<pre class="ql-syntax" spellcheck="false">with fits.open(driz_dir+'omega_cen_f606w_drz.fits') as hdul:
    sci_driz = np.nan_to_num(hdul['SCI'].data)
print...
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/hst_notebooks/pull/270/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='70'/>

haticekaratay commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1551 @@
+{
```
<p>Please add a descriptive alt text. This ensures that the images is accessible to users who rely on screen readers.</p><p> </p>

<pre class="ql-syntax" spellcheck="false">![your description](workflow.png)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/hst_notebooks/pull/270/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

mrevalski replied:
I've added a fairly extensive alt text description, which is included below for reference. Please let me know if this is insufficient:

_This flowchart provides users with a general guide for deciding which PSF modeling technique is the most relevant based on their dataset and science goals. The top of the chart asks users if they are modeling at the exposure or drizzle level. The drizzle workflow forks to the right, and asks users how many bright, isolated stars are in their image. If there are fewer than 10 stars, they can fit an analytical profile or download custom tools as described below. If they have 10 or more stars, they can extract and stack stars as described in Section 3. If users are working at the exposure level, then the diagram forks to the left. If the user's goal is focused on high-accuracy photometry and astrometry of stars, then they follow the left fork. If their goals are to model the extended wings and diffraction spikes of the PSF, they follow the fork right. In following the left fork for precision photometry and astrometry, users are asked if an empirical model from the WFC3 team exists. If it does, then they can run hst1pass and examine the residuals. If there is not an empirical model, they can use the model for the closest filter and perturb it with hst1pass if there are at least 10 stars. In following the right fork for modeling extended wing emission and diffraction spikes, users are asked if they have more or less than 10 stars in their image. If they have fewer than 10, they can use the MAST cutout service to retrieve and stack stars from other observations. If they have at least 10 stars, then they can stack stars directly from their observations. This flowchart is a general guide, and we note that several different paths may be applicable to a single dataset. Users are encouraged to experiment with the options below._

haticekaratay commented on notebooks/WFC3/point_spread_function/hst_point_spread_function.ipynb:
```diff
@@ -0,0 +1,1551 @@
+{
```
<p>A small improvement can be made for readability. Instead of using <code>if os.path.exists(code_dir+'hst1pass.e') is False</code>  it's more readable to use </p><p><code>if not os.path.exists(code_dir+'hst1pass.e'):</code></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/spacetelescope/hst_notebooks/pull/270/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

mrevalski replied:
Personally I find the existing line most readable, as it makes it clear that `os.path.exists` returns a boolean value.

I would like to retain this line, but if you disagree for best coding practices please let me know and I'll change it?

haticekaratay commented:
Your notebook is exceptionally well-crafted and comprehensive. The introduction is clear and informative, and the links to tutorials and additional resources are incredibly helpful. The way you've linked every imported package and the workflow is exemplary. This is one of the best notebooks I've read so far.

mrevalski commented:
> Your notebook is exceptionally well-crafted and comprehensive. The introduction is clear and informative, and the links to tutorials and additional resources are incredibly helpful. The way you've linked every imported package and the workflow is exemplary. This is one of the best notebooks I've read so far.

Hi @haticekaratay, thank you for the helpful review and very, very kind words! While I invested a lot of effort into this notebook, I owe much credit to @FDauphin and the WFC3 team for the helpful suggestions. I have implemented all of your changes, and also made some minor tweaks to improve the wording for a few descriptions in the markdown, and the notebook is now ready for your review. Please let me know if there are any additional polishing touches that are required before this can be merged!
